langchain/src/chat_models/baiduernie.ts

```diff
@@ -0,0 +1,438 @@
+            }
+          ).catch((error) => {
+            if (!rejected) {
+              rejected = true;
+              reject(error);
+            }
+          });
+        })
+      : await this.completionWithRetry(
+          {
+            ...params,
+            messages: messagesMapped,
+          },
+          {
+            signal: options?.signal,
+          }
+        );
+
+    const {
+      completion_tokens: completionTokens,
+      prompt_tokens: promptTokens,
+      total_tokens: totalTokens,
+    } = data.usage ?? {};
+
+    if (completionTokens) {
+      tokenUsage.completionTokens =
+        (tokenUsage.completionTokens ?? 0) + completionTokens;
+    }
+
+    if (promptTokens) {
+      tokenUsage.promptTokens = (tokenUsage.promptTokens ?? 0) + promptTokens;
+    }
+
+    if (totalTokens) {
+      tokenUsage.totalTokens = (tokenUsage.totalTokens ?? 0) + totalTokens;
+    }
+
+    const generations: ChatGeneration[] = [];
+    const text = data.result ?? "";
+    generations.push({
+      text,
+      message: new AIChatMessage(text),
+    });
+    return {
+      generations,
+      llmOutput: { tokenUsage },
+    };
+  }
+
+  /** @ignore */
+  async completionWithRetry(
+    request: ChatCompletionRequest,
+    options?: StreamingAxiosConfiguration
+  ) {
+    // The first run will get the accessToken
+    if (!this.accessToken) {
+      this.accessToken = await this.getAccessToken();
+    }
+
+    const axiosOptions = {
+      params: { access_token: this.accessToken },
+      headers: { "Content-Type": "application/json" },
+      adapter: isNode() ? undefined : fetchAdapter,
+      ...options,
+    } as StreamingAxiosConfiguration;
+
+    const makeCompletionRequest = async () => {
+      const res = await axios.post(this.apiUrl, request, axiosOptions);
+      return res.data;
+    };
+    return this.caller.call(makeCompletionRequest);
+  }
+
+  _llmType() {
+    return "ernie";
+  }
+
+  /** @ignore */
+  _combineLLMOutput() {
+    return [];
+  }
+}
+
+export class ChatBaiduERNIE
+  extends ChatBaiduERNIEBase
+  implements BaiduERNIEChatInput
+{
+  temperature?: number | undefined;
+
+  topP?: number | undefined;
+
+  penaltyScore?: number | undefined;
+
+  constructor(fields?: Partial<BaiduERNIEChatInput> & BaseChatModelParams) {
+    super(fields);
+
+    this.apiUrl =
+      "https://aip.baidubce.com/rpc/2.0/ai_custom/v1/wenxinworkshop/chat/completions";
+
+    this.temperature = fields?.temperature ?? this.temperature;
+    this.topP = fields?.topP ?? this.topP;
+    this.penaltyScore = fields?.penaltyScore ?? this.penaltyScore;
+
+    this.modelName = "ERNIE-Bot";
+  }
+
+  /**
+   * Get the parameters used to invoke the model
+   */
+  invocationParams(): Omit<ChatCompletionRequest, "messages"> {
+    return {
+      ...super.invocationParams(),
+      temperature: this.temperature,
+      top_p: this.topP,
+      penalty_score: this.penaltyScore,
+    };
+  }
+}
+
+export class ChatBaiduERNIETurbo
```

Comment:
We've generally been splitting models by provider and using a `modelName` parameter to distinguish between things like `turbo` vs. standard - can you change to that?

On the same note, I think we can call this `ChatBaidu` instead of `ChatBaiduErnie` and just have that as the model name.

Reply:
Yes, I understand your request. But let me explain why I made this weird choice.

Baidu provides different APIs for [ERNIE](https://cloud.baidu.com/doc/WENXINWORKSHOP/s/jlil56u11) and [ERNIE-turbo](https://cloud.baidu.com/doc/WENXINWORKSHOP/s/4lilb2lpf). Most of them are the same, but endpoints and some parameters are different. For example, ERNIE has temperature, but ERNIE-turbo has not.

I was afraid that they will make more differences, so two classes maybe are better choice. But just as I'm typing these words, I get a info from a Baiduer, they will merge the API finally. So no problem to merge them as one class. I will do it.

The second question: `ChatBaidu`, `ChatBaiduErnie` or `ChatErnie`?

I followed the `ChatGoogleVertexAI` since Baidu is like Google in China. But `ChatBaiduWenxin` is more accuracy because Wenxin is the brand name of Baidu LLM-based AI cloud services just as Vertex AI.

How about using `ChatBaiduWenxin`?

Reply:
Yeah if some model interfaces are volatile let's add support for them when they stabilize and start with the most stable interface first.

ChatBaiduWenxin sounds good to me!